### PR TITLE
Issue/14881 closures for navigation entries

### LINKED ISSUE #14881: Do not create navigation entries unless required
### Worst Case, aka Current State
Currently apps register their navigation entry with an array in app.php e.g. Activity:
```php
$app = new Application();
$c = $app->getContainer();

$navigationEntry = array(
	'id' => $c->getAppName(),
	'order' => 1,
	'name' => $c->query('ActivityL10N')->t('Activity'),
	'href' => \OC::$server->getURLGenerator()->linkToRoute('activity.Activities.showList'),
	'icon' => \OC::$server->getURLGenerator()->imagePath('activity', 'activity.svg'),
);
\OC::$server->getNavigationManager()->add($navigationEntry);
```

This creates a lot of overhead, although not required:

1. The `Application` of the activity app is created and instanciated
2. The `$app->getContainer()` ....
3. The `$c->query('ActivityL10N')` ....
4. The `\OC::$server->getURLGenerator()` ....

This makes requests that do not need the navigation quite bloated.

I got two ideas to prevent this easily:

---

### 1. Use Closures

Add a `addClosure()` method and register all navigation per closure, so the stuff is only created when required, on `getAll()` of the NavigationManager run the closures and add them:

    ```php
    	public function getAll() {
    		foreach ($this->closureEntries as $c) {
    			$this->add($c());
    		}
    		$this->closureEntries = array();
    		return $this->entries;
    	}
    ```

App Code:
```php
\OC::$server->getNavigationManager()->addClosure(function () use ($c) {
	$app = new Application();
	$c = $app->getContainer();
	$urlGenerator = $c->getServer()->getURLGenerator();

	return [
		'id' => $c->getAppName(),
		'order' => 1,
		'name' => $c->query('ActivityL10N')->t('Activity'),
		'href' => $urlGenerator->linkToRoute('activity.Activities.showList'),
		'icon' => $urlGenerator->imagePath('activity', 'activity.svg'),
	];
});
```
#### Advantage:
+ Code is not executed unless really required
+ Quick (and dirty)

#### Disadvantage:
- We create tones of closures
- Return value of the closure is not specific and needs to be verified?
- Apps need to be updated to help improve the performance

---

### 2. Use a method in the appframework
Add a `registerNavigation()` method to the `\OCP\AppFramework\App` class, and call that for all enabled apps in the `NavigationManager::getAll()` method.

#### Advantage:
+ Code is not executed unless really required
+ Properly structured and making use of the Application class

#### Disadvantage:
- We need a way to create a list of all `\OCP\AppFramework\App` objects, so we can run the `registerNavigation()` method on them (but I think we create all those classes already anyway, and just don't keep them in the appmanager or whereever)
- Apps need to be updated to help improve the performance

---

Both ways are way finer and less ressource wasting then the current implementation.

Opinions? @LukasReschke @Raydiation @DeepDiver1975

### PR DESCRIPTION
**Pull request by nickvergessen**
Speeds up app.php by `0.04s` per app that uses a closure

Fix #14881

### CODE REVIEW

**Comment by nickvergessen** [comment]
Sample patch on the activity app: https://github.com/owncloud/activity/pull/261/files?w=1

@DeepDiver1975 @Raydiation as discussed

**Comment by MorrisJobke** [comment]
> 0.04

Unit? Hours would be nice. Kilometer would be weird and m/s would be strange. ;)

**Comment by Raydiation** [comment]
@MorrisJobke apparently seconds which amounts to 0.5 seconds for 10 apps which is very common :)

File: lib/private/navigationmanager.php [hunk 1]
```diff
@@ -14,13 +14,23 @@
...
-	 * @param array $entry containing: id, name, order, icon and href key
+	 *
+	 * @param array|\Closure $entry Array containing: id, name, order, icon and href key
+	 *					The use of a closure is preferred, because it will avoid
+	 * 					loading the routing of your app, unless required.
+	 * @return void
 	 */
-	public function add(array $entry) {
+	public function add($entry) {
+		if ($entry instanceof \Closure) {
```
Comment (Raydiation) [inline lib/private/navigationmanager.php]: I'd just append it to the entries like normal and not use a separate array. That way you can also defer that to the point when its actually used

Comment (Raydiation) [inline lib/private/navigationmanager.php]: Ah right, makes more sense from a caching POV, nvm :+1:

**Comment by Raydiation** [comment]
:+1: not tested but implementation looks fine

**Comment by LukasReschke** [comment]
:+1:

**Comment by MorrisJobke** [comment]
@LukasReschke @nickvergessen Please at this to the 8.1 feature page ;)

**Comment by MorrisJobke** [comment]
And update the docs ;)

**Comment by nickvergessen** [comment]
- [x] Mailing list: https://mailman.owncloud.org/pipermail/devel/2015-March/001146.html
- [x] Feature page: https://github.com/owncloud/core/wiki/ownCloud-8.1-Features#performance
- [x] ocdev 8.1 app template: https://github.com/owncloud/ocdev/pull/30/files
- [x] Docs: https://github.com/owncloud/documentation/pull/961

**Comment by MorrisJobke** [comment]
@nickvergessen :+1:

**Pull request merged**